### ISSUE TITLE
segfault in `check_antenna`

### ISSUE THREAD

**Issue by gadfort** [The-OpenROAD-Project/OpenROAD#3803]
### Describe the bug

When running `check_antenna` OpenROAD segfaults.

```
 0# 0x000055C036B030AB in openroad
 1# 0x00007F98FDBB2520 in /lib/x86_64-linux-gnu/libc.so.6
 2# odb::dbWireEncoder::addITerm(odb::dbITerm*) in openroad
 3# odb::tmg_conn::addToWire(int, int, int, bool, bool) in openroad
 4# odb::tmg_conn::treeReorder(bool) in openroad
 5# odb::tmg_conn::analyzeNet(odb::dbNet*) in openroad
 6# grt::RepairAntennas::makeNetWire(odb::dbNet*, std::vector<grt::GSegment, std::allocator<grt::GSegment> >&, std::map<int, odb::dbTechVia*, std::less<int>, std::allocator<std::pair<int const, odb::dbTechVia*> > >&) in openroad
 7# grt::RepairAntennas::makeNetWires(std::map<odb::dbNet*, std::vector<grt::GSegment, std::allocator<grt::GSegment> >, grt::cmpById, std::allocator<std::pair<odb::dbNet* const, std::vector<grt::GSegment, std::allocator<grt::GSegment> > > > >&, int) in openroad
 8# ant::AntennaChecker::checkAntennas(odb::dbNet*, bool) in openroad
 9# ant::check_antennas(char const*, bool) in openroad
10# 0x000055C0376BABD4 in openroad
```

### Expected Behavior

No segfault

### Environment

```shell
kernel: Linux 5.15.0-1040-azure
os: Ubuntu 22.04.2 LTS (Jammy Jellyfish)
cmake version 3.26.4
-- The CXX compiler identification is GNU 11.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- OpenROAD version: v2.0-9702-g92f8e07b0
-- System name: Linux
-- Compiler: GNU 11.4.0
-- Build type: RELEASE
-- Install prefix: /usr/local
-- C++ Standard: 17
-- C++ Standard Required: ON
-- C++ Extensions: OFF
-- The C compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Python: /usr/bin/python3.10 (found version "3.10.12") found components: Interpreter 
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Performing Test C_COMPILER_SUPPORTS__-Wall
-- Performing Test C_COMPILER_SUPPORTS__-Wall - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wall
-- Performing Test CXX_COMPILER_SUPPORTS__-Wall - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-array-bounds
-- Performing Test C_COMPILER_SUPPORTS__-Wno-array-bounds - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-array-bounds
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-array-bounds - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-nonnull
-- Performing Test C_COMPILER_SUPPORTS__-Wno-nonnull - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-nonnull
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-nonnull - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-maybe-uninitialized
-- Performing Test C_COMPILER_SUPPORTS__-Wno-maybe-uninitialized - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-maybe-uninitialized
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-maybe-uninitialized - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format-overflow
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format-overflow - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format-overflow
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format-overflow - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-variable
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-variable - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-variable
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-variable - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-function
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-function - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-function
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-function - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-write-strings
-- Performing Test C_COMPILER_SUPPORTS__-Wno-write-strings - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-write-strings
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-write-strings - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-sign-compare
-- Performing Test C_COMPILER_SUPPORTS__-Wno-sign-compare - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-sign-compare
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-sign-compare - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-deprecated
-- Performing Test C_COMPILER_SUPPORTS__-Wno-deprecated - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-deprecated
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-deprecated - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-c++11-narrowing
-- Performing Test C_COMPILER_SUPPORTS__-Wno-c++11-narrowing - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-c++11-narrowing
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-c++11-narrowing - Failed
-- Performing Test C_COMPILER_SUPPORTS__-Wno-register
-- Performing Test C_COMPILER_SUPPORTS__-Wno-register - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-register
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-register - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format
-- Performing Test C_COMPILER_SUPPORTS__-Wno-format - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-format - Success
-- Performing Test C_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal
-- Performing Test C_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-reserved-user-defined-literal - Failed
-- Performing Test C_COMPILER_SUPPORTS__-fpermissive
-- Performing Test C_COMPILER_SUPPORTS__-fpermissive - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-fpermissive
-- Performing Test CXX_COMPILER_SUPPORTS__-fpermissive - Success
-- Performing Test C_COMPILER_SUPPORTS__-x
-- Performing Test C_COMPILER_SUPPORTS__-x - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__-x
-- Performing Test CXX_COMPILER_SUPPORTS__-x - Failed
-- Performing Test C_COMPILER_SUPPORTS__c++
-- Performing Test C_COMPILER_SUPPORTS__c++ - Failed
-- Performing Test CXX_COMPILER_SUPPORTS__c++
-- Performing Test CXX_COMPILER_SUPPORTS__c++ - Failed
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-but-set-variable
-- Performing Test C_COMPILER_SUPPORTS__-Wno-unused-but-set-variable - Success
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-but-set-variable
-- Performing Test CXX_COMPILER_SUPPORTS__-Wno-unused-but-set-variable - Success
-- TCL library: /usr/lib/x86_64-linux-gnu/libtcl.so
-- TCL header: /usr/include/tcl/tcl.h
-- TCL readline library: /usr/lib/x86_64-linux-gnu/libtclreadline.so
-- TCL readline header: /usr/include/x86_64-linux-gnu
-- Found SWIG: /usr/bin/swig4.0 (found suitable version "4.0.2", minimum required is "3.0")  
-- Found Boost: /usr/local/lib/cmake/Boost-1.80.0/BoostConfig.cmake (found version "1.80.0")  
-- boost: 1.80.0
-- Found Python3: /usr/include/python3.10 (found version "3.10.12") found components: Development Development.Module Development.Embed 
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")  
-- spdlog: 1.8.1
-- Found BISON: /usr/bin/bison (found version "3.8.2") 
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- STA version: 2.4.0
-- STA git sha: 7cf916ba205115a06c4531a044ced481f1ff8f12
-- System name: Linux
-- Compiler: GNU 11.4.0
-- Build type: RELEASE
-- Build CXX_FLAGS: -O3 -DNDEBUG
-- Install prefix: /usr/local
-- Found FLEX: /usr/bin/flex (found version "2.6.4") 
-- TCL library: /usr/lib/x86_64-linux-gnu/libtcl.so
-- TCL header: /usr/include/tcl/tcl.h
-- SSTA: 0
-- STA executable: /home/gadfort/OpenROAD/src/sta/app/sta
-- GPU is not enabled
-- TCL library: /usr/lib/x86_64-linux-gnu/libtcl.so
-- TCL header: /usr/include/tcl/tcl.h
-- Found re2: /opt/or-tools/lib/cmake/re2/re2Config.cmake (found version "9.0.0") 
-- Found Clp: /opt/or-tools/lib/cmake/Clp/ClpConfig.cmake (found version "1.17.7") 
-- Found Cbc: /opt/or-tools/lib/cmake/Cbc/CbcConfig.cmake (found version "2.10.7") 
-- Found Eigen3: /usr/local/share/eigen3/cmake/Eigen3Config.cmake (found version "3.4.0") 
-- Found SCIP: /opt/or-tools/lib/cmake/scip/scip-config.cmake (found version "8.0.1") 
-- GUI is enabled
-- Charts widget is not enabled
-- Found Boost: /usr/local/lib/cmake/Boost-1.80.0/BoostConfig.cmake (found version "1.80.0") found components: serialization 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Could NOT find VTune (missing: VTune_LIBRARIES VTune_INCLUDE_DIRS) 
-- Found Boost: /usr/local/lib/cmake/Boost-1.80.0/BoostConfig.cmake (found suitable version "1.80.0", minimum required is "1.78")  
-- TCL library: /usr/lib/x86_64-linux-gnu/libtcl.so
-- TCL header: /usr/include/tcl/tcl.h
-- Found Boost: /usr/local/lib/cmake/Boost-1.80.0/BoostConfig.cmake (found version "1.80.0") found components: serialization system thread 
-- TCL readline enabled
-- Tcl Extended disabled
-- Python3 enabled
-- Configuring done (4.8s)
-- Generating done (0.7s)
```

### To Reproduce

[sc_issue_ibex_job0_freepdk45_demo_route0_20230806-170640.tar.gz](https://github.com/The-OpenROAD-Project/OpenROAD/files/12270064/sc_issue_ibex_job0_freepdk45_demo_route0_20230806-170640.tar.gz)

```
tar xvf sc_issue_ibex_job0_freepdk45_demo_route0_20230806-170640.tar.gz
cd sc_issue_ibex_job0_freepdk45_demo_route0_20230806-170640
./run.sh
```

### Relevant log output

```shell
[INFO GRT-0043] No OR_DEFAULT vias defined.
Signal 11 received
Stack trace:

 0# 0x000055C036B030AB in openroad
 1# 0x00007F98FDBB2520 in /lib/x86_64-linux-gnu/libc.so.6
 2# odb::dbWireEncoder::addITerm(odb::dbITerm*) in openroad
 3# odb::tmg_conn::addToWire(int, int, int, bool, bool) in openroad
 4# odb::tmg_conn::treeReorder(bool) in openroad
 5# odb::tmg_conn::analyzeNet(odb::dbNet*) in openroad
 6# grt::RepairAntennas::makeNetWire(odb::dbNet*, std::vector<grt::GSegment, std::allocator<grt::GSegment> >&, std::map<int, odb::dbTechVia*, std::less<int>, std::allocator<std::pair<int const, odb::dbTechVia*> > >&) in openroad
 7# grt::RepairAntennas::makeNetWires(std::map<odb::dbNet*, std::vector<grt::GSegment, std::allocator<grt::GSegment> >, grt::cmpById, std::allocator<std::pair<odb::dbNet* const, std::vector<grt::GSegment, std::allocator<grt::GSegment> > > > >&, int) in openroad
 8# ant::AntennaChecker::checkAntennas(odb::dbNet*, bool) in openroad
 9# ant::check_antennas(char const*, bool) in openroad
10# 0x000055C0376BABD4 in openroad
11# TclNRRunCallbacks in /lib/x86_64-linux-gnu/libtcl8.6.so
12# 0x00007F9901A4A028 in /lib/x86_64-linux-gnu/libtcl8.6.so
13# Tcl_EvalEx in /lib/x86_64-linux-gnu/libtcl8.6.so
14# Tcl_Eval in /lib/x86_64-linux-gnu/libtcl8.6.so
15# sta::sourceTclFile(char const*, bool, bool, Tcl_Interp*) in openroad
16# ord::tclAppInit(Tcl_Interp*) in openroad
17# Tcl_MainEx in /lib/x86_64-linux-gnu/libtcl8.6.so
18# main in openroad
19# 0x00007F98FDB99D90 in /lib/x86_64-linux-gnu/libc.so.6
20# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
21# _start in openroad
```

**Comment by maliberty** [comment]
I find these SiC packaged test cases hard to follow.  How do I determine exactly what files/commands are being executed?

**Comment by gadfort** [comment]
@maliberty 
The openroad scripts are located in: build/ibex/job0_freepdk45_demo/sc_collected_files/scripts_fccbfa790db5ddcc0abfcccac3b21bc686d7535a/

All the files needed to run this are placed in build/ibex/job0_freepdk45_demo/sc_collected_files

The run.sh script is simply going to call:
```
./build/ibex/job0_freepdk45_demo/route/0/replay.sh
```
which is the script that executes OpenROAD

The two files are going to be: `sc_apr.tcl` and `sc_route.tcl`

**Comment by maliberty** [comment]
@eder-matheus what is your status?

**Comment by eder-matheus** [comment]
@maliberty I'm trying to isolate the problem, to understand on which net the crash happens and how the wires for it are, but I'm having trouble reproducing the crash. Sometimes it just goes through the routing flow without errors, but I'm also having this error:
<img width="958" alt="image" src="https://github.com/The-OpenROAD-Project/OpenROAD/assets/43450810/0f868e12-4edf-40bb-a271-11639bf597e6">

I'm running in a MacOS, with the latest openroad compiled in debug mode. I'll try it on CentOS too.

**Comment by maliberty** [comment]
Please try with valgrind

**Comment by gadfort** [comment]
@eder-matheus the stuck issue seems to be from the boost stracktrace getting into a deadlock with a atomic lock

**Comment by gadfort** [comment]
@eder-matheus I've narrowed it down a little.
If I exclude the net `rst_ni` from checking and from wire creation, doesn't segfault. It looks like this net has a fanout of 1700 (it's a port), so the wire that GRT is encoding is probably corrupted (so there probably is a bug in the GRT wire encoding code).

This seems to be from not having a driving cell set on the inputs or a max_fanout causing the `repair_design` to ignore the net. @maliberty I can fix the constraints since they are wrong, but to seems like there should be some kind of warning either in `check_setup` or `repair_design`? Not sure what the trigger should be, but no cell in the library can drive this kind of fanout.

**Comment by eder-matheus** [comment]
> @eder-matheus I've narrowed it down a little. If I exclude the net `rst_ni` from checking and from wire creation, doesn't segfault. It looks like this net has a fanout of 1700 (it's a port), so the wire that GRT is encoding is probably corrupted (so there probably is a bug in the GRT wire encoding code).
> 
> This seems to be from not having a driving cell set on the inputs or a max_fanout causing the `repair_design` to ignore the net. @maliberty I can fix the constraints since they are wrong, but to seems like there should be some kind of warning either in `check_setup` or `repair_design`? Not sure what the trigger should be, but no cell in the library can drive this kind of fanout.

Thanks for the update, @gadfort. The main issue is how we evaluate the global routing to search for antenna violations. The GRT tool does not create local connections with the pins, just the connection between the guides. When we evaluate the antennas, we convert the guides to wires and create the connection with the pins. In these connections, we can have overlaps between the wires and pin shapes, leading to the crash in order_wires.

We already made some improvements on this conversion process, trying to avoid the overlaps, but it may be hard in a scenario like yours, with a high fanout net.

I'm wondering if we should change this approach and, instead of converting the groute to actual wires, make the antenna checker understand the route format and evaluate it.

For now, I'll dig at this net you mentioned to understand precisely what issue order_wires is having trouble with.

**Comment by gadfort** [comment]
@maliberty thats a good point, I thought it was turned on in that run, but certainly didn't seem to occur. But I would still advocate for some warnings if that is possible.

**Comment by gadfort** [comment]
yeah, `check_setup` would be ideal, 
Warn on:
1. set_driving_cell or set_max_fanout on input_ports
2. set_load on output_ports
I'm not sure what the right combinations would be, but I wouldn't generate a warning if driving cell is set and max_fanout is not.

**Comment by maliberty** [comment]
Please open an enhancement request for this. 

For (1) I wouldn't warn if you have max cap/slew set as those effectively limit the fanout as well.

**Comment by gadfort** [comment]
Done https://github.com/The-OpenROAD-Project/OpenROAD/issues/3884

**Comment by maliberty** [comment]
@eder-matheus was this crash fixed?
